src/lib/operators/insert.cpp

```diff
@@ -107,136 +93,132 @@ std::shared_ptr<const Table> Insert::_on_execute(std::shared_ptr<TransactionCont
       for (ColumnID column_id{0}; column_id < target_chunk->column_count(); ++column_id) {
-        const auto& source_segment = source_chunk->get_segment(column_id);
-        typed_segment_processors[column_id]->copy_data(source_segment, source_chunk_start_index,
-                                                       target_chunk->get_segment(column_id), target_start_index,
-                                                       num_to_insert);
+        const auto source_segment = source_chunk->get_segment(column_id);
+        const auto target_segment = target_chunk->get_segment(column_id);
+
+        resolve_data_type(_target_table->column_data_type(column_id), [&](const auto data_type_t) {
+          using ColumnDataType = typename decltype(data_type_t)::type;
+          copy<ColumnDataType>(source_segment, source_row_id.chunk_offset, target_segment, target_chunk_offset,
+                               num_rows_current_iteration);
+        });
       }
-      still_to_insert -= num_to_insert;
-      target_start_index += num_to_insert;
-      source_chunk_start_index += num_to_insert;
-
-      bool source_chunk_depleted = source_chunk_start_index == source_chunk->size();
-      if (source_chunk_depleted) {
-        source_chunk_id++;
-        source_chunk_start_index = 0u;
+
+      if (num_rows_current_iteration == source_chunk_remaining_rows) {
+        // Proceed to next source Chunk
+        ++source_row_id.chunk_id;
+        source_row_id.chunk_offset = 0;
+      } else {
+        source_row_id.chunk_offset += num_rows_current_iteration;
       }
-    }
 
-    for (auto i = start_index; i < start_index + current_num_rows_to_insert; i++) {
-      // we do not need to check whether other operators have locked the rows, we have just created them
-      // and they are not visible for other operators.
-      // the transaction IDs are set here and not during the resize, because
-      // tbb::concurrent_vector::grow_to_at_least(n, t)" does not work with atomics, since their copy constructor is
-      // deleted.
-      target_chunk->get_scoped_mvcc_data_lock()->tids[i] = context->transaction_id();
-      _inserted_rows.emplace_back(RowID{target_chunk_id, i});
+      target_chunk_offset += num_rows_current_iteration;
+      target_chunk_range_remaining_rows -= num_rows_current_iteration;
     }
-
-    input_offset += current_num_rows_to_insert;
-    start_index = 0u;
   }
 
   return nullptr;
 }
 
 void Insert::_on_commit_records(const CommitID cid) {
-  for (auto row_id : _inserted_rows) {
-    auto chunk = _target_table->get_chunk(row_id.chunk_id);
-
-    auto mvcc_data = chunk->get_scoped_mvcc_data_lock();
-    mvcc_data->begin_cids[row_id.chunk_offset] = cid;
-    mvcc_data->tids[row_id.chunk_offset] = 0u;
+  for (const auto& target_chunk_range : _target_chunk_ranges) {
+    const auto target_chunk = _target_table->get_chunk(target_chunk_range.chunk_id);
+    auto mvcc_data = target_chunk->get_scoped_mvcc_data_lock();
```

Comment:
Do we need to hold the lock for the entire inner loop? (below too)

Reply:
tbh I don't even know why we need that lock. The MvccData uses concurrent_vectors, after all, and Insert holds a lock over the entire table when it resizes them...

Reply:
I believe it has something to do with the MVCC data being migrated if a chunk is being encoded. Will have to look into it again, though.

Reply:
It is needed when we shrink the MVCC data:

```
  /**
   * Compacts the internal representation of
   * the mvcc data in order to reduce fragmentation
   * Locks mvcc data exclusively in order to do so
   */
  void shrink();
```

However, 9b15cafcc98e8f82ae731e7601a2fa701ee047bb removed the exclusive lock by mistake. It should be reintroduced (and then removed again by #749).

Reply:
done